### PR TITLE
 Fix Select-Object to approve UX and fix return a property named *

### LINKED ISSUE #2351: Select-Object -ExcludeProperty dependency on -Property is unintuitive.
<!--

If it is a bug report:
- make sure you are able to repro it on the latest released version. 
You can install the latest version from https://github.com/PowerShell/PowerShell/releases
- Search the existing issues.
- Refer to the [FAQ](../docs/FAQ.md).
- Refer to the [known issues](../docs/KNOWNISSUES.md).
- Fill out the following repro template

If it's not a bug, please remove the template and elaborate the issue in your own words.
-->
## Steps to reproduce

`Get-Service -Name Spooler | Select-Object -ExcludeProperty Status`
## Expected behavior

Return a Selected.System.ServiceProcess.ServiceController object with all standard properties of System.ServiceProcess.ServiceController except Status.
## Actual behavior

Returns System.ServiceProcess.ServiceController object with all properties including status. 

To get the desired behavior you must use the -Property parameter in conjunction with the -ExcludeProperty parameter, this is explained in the help documentation for select-object as: 

"This parameter is effective only when the command also includes the Property parameter."

I don't understand the case where you would want to declare to explicitly include a property, then exclude it again such as: 

`Get-Service -Name Spooler | Select-Object -Property Status -ExcludeProperty Status`

Therefore the convention is to always pass "-Property *" when using the -ExcludeProperty parameter. I would suggest defaulting to -Property \* when using the -ExcludeProperty cmdlet, if that would be considered a breaking change, I would suggest throwing an error if -ExcludeProperty is used without -Property indicating that you are missing a mandatory parameter since ExcludeProperty is dependent on -Property. 
## Environment data

<!-- provide the output of $PSVersionTable -->

``` powershell
> $PSVersionTable
Name                           Value                                                                                                                                                                                                                                                                                                                      
----                           -----                                                                                                                                                                                                                                                                                                                      
PSVersion                      5.0.10586.63                                                                                                                                                                                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                                                                                                                    
BuildVersion                   10.0.10586.63                                                                                                                                                                                                                                                                                                              
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                                                                                                            
WSManStackVersion              3.0                                                                                                                                                                                                                                                                                                                        
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                                                                                                                        
SerializationVersion           1.1.0.1   
```

### LINKED ISSUE #2420: Select-Object return a property named *
## Steps to reproduce

[pscustomobject]@{Thing="thing1"} | Select-Object -Property \* -ExcludeProperty thing 
## Expected behavior

The cmdlet return empty pscustomobject (w/o properties).
## Actual behavior

The cmdlet return:

```
 * 
 - 
```
## Environment data

``` powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.0.10240.17113
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   10.0.10011.16384
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
```

### PR DESCRIPTION
**Pull request by iSazonov**
1. Now ExcludeProperty imply Property = '*' to approve UX #2351
2. Don't return property named '*'  #2420

Close #2351 
Close #2420

### CODE REVIEW

**Comment by iSazonov** [comment]
@PowerShellTeam please review the code.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs [hunk 1]
```diff
@@ -414,7 +419,7 @@ private void ProcessParameter(MshParameter p, PSObject inputObject, List<PSNoteP
                 }
             }
 
-            if (expressionResults.Count == 0)
+            if (expressionResults.Count == 0 && !ex.HasWildCardCharacters)
             {
                 //Commented out for bug 1107600
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Could you please remove the commented-out code snippet here? It's useless and there is no way to figure out what `bug 1107600` is now :)

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: And it would be better if you can add some comments here, explaining your purpose (allow `Select-Object -Property noexist-name` to return a `PSObject` with property `noexist-name`, unless `noexist-name` itself contains wildcards).
I find the implementation of `Select-Object` is so poorly commented (near to zero comment) that it becomes a pain to read the code. So let's add some when we make changes in this file.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 2]
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
 		$results[0] | Should Be "3"
 	}
 }
+
+Describe "Select-Object with Property = '*'" -Tags "CI" {
+
+        # Issue #2420
+	It "Select-Object with implicit Property = '*' don't return property named '*'"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -ExcludeProperty thing 
+		$results.psobject.Properties.Item("*").Name | Should Be $null
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: > Item("*").Name

`.Name` seems not necessary here.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs [hunk 3]
```diff
@@ -337,6 +337,11 @@ private void ProcessExpressionParameter()
             if (ExcludeProperty != null)
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                if ((Property == null) || (Property.Length == 0))
+                {
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: This will result a breaking change -- the scirpt below runs successfully before this change; but with this change a bunch errors will be thrown out.
```powershell
$s = Get-Process -Id $pid | Select-Object -ExcludeProperty ProcessName -ExpandProperty Modules
```
However, I don't feel this would break any existing scripts practically. It's documented that `-ExcludeProperty` has no effect without `-Property`, so I don't think any existing script would only specify `-ExcludeProperty`.

Maybe we can analyze the powershell scripts on GitHub to get some data. We can borrow @lzybkr's [AnalyzeCurlUsage.ps1](https://gist.github.com/lzybkr/3cd091334355f381d1d6ee7acfad5a48).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 4]
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+	}
+
+        # Issue #2420
+	It "Select-Object with explicit Property = '*' don't return property named '*'"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -Property * -ExcludeProperty thing 
+		$results.psobject.Properties.Item("*").Name | Should Be $null
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: `.Name` seems not necessary.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 5]
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+	It "Select-Object with explicit Property = '*' exclude single property"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -Property * -ExcludeProperty thing
+		$results.psobject.Properties.Item("Thing") | Should Be $null
+		$results.psobject.Properties.Item("*") | Should Be $null
+	}
+
+        # Issue #2351
+	It "Select-Object with implicit Property = '*' exclude not single prperty"{
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -ExcludeProperty Param2 
+                $results.Param2 | Should Be $null
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Please fix the indentation here.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 6]
```diff
@@ -248,3 +248,46 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+		$results.Thing | Should Be "thing1"
+	}
+
+        # Issue #2351
+	It "Select-Object with explicit Property = '*' exclude not single prperty"{
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -Property * -ExcludeProperty Param2 
+                $results.Param2 | Should Be $null
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Please fix the indentation here.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: 1.The current tests do not catch it. I think we should add this in tests.

2.There is no difference between:
```powershell
 $s = Get-Process -Id $pid | Select-Object -ExpandProperty Modules
$s[0] | fl *
```
and
```powershell
$s = Get-Process -Id $pid | Select-Object -ExcludeProperty FileName -ExpandProperty Modules
$s[0] | fl *
```
i.e.  ExpandProperty skip ExcludeProperty. So the change could be corrected without  breaking.

3.The cmdlet is very intuitive to the user UX and very controversial:
ExcludeProperty is skipped w/o Property
ExpandProperty skip ExcludeProperty
ExpandProperty with Property throw exception
Compare:
```powershell
1 | Select-Object
"1" | Select-Object
1 | Select-Object -Property *
"1" | Select-Object -Property *
```
and many more.
And it is a tricky code, and any even slightly change this code can easily break down. 

(In general I think this cmdlet should be replaced with a newer, which meets modern requirements to cmdlets.)

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 1 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: I generally see little comments in code and tests and thought it was best practice. I will make more comments.
I think @PowerShellTeam  needs to mark this moment in documentation: "make more comments in your code".

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (Jaykul) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: I think you mean that it's very **un**intuitive 😕 

If the intention is (to maintain the current expected behavior) that `-ExpandProperty` ignores `-ExcludeProperty` wouldn't it be better to have these switches in separate parameter sets? I'd certainly rather have it be _obvious that they can't be used together_ than to allow them together but ignore one (or worse, produce runtime errors if they're used together)...

**Comment by RichardSiddaway** [comment]
Having separate parameter sets would be a great idea

Sent from my iPad

On 9 Oct 2016, at 20:30, Joel Bennett <notifications@github.com<mailto:notifications@github.com>> wrote:


@Jaykul commented on this pull request.

________________________________

In src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs<https://github.com/PowerShell/PowerShell/pull/2421>:

> @@ -337,6 +337,11 @@ private void ProcessExpressionParameter()
             if (ExcludeProperty != null)
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                if ((Property == null) || (Property.Length == 0))
+                {


I think you mean that it's very unintuitive ?

If the intention is (to maintain the current expected behavior) that -ExpandProperty ignores -ExcludeProperty wouldn't it be better to have these switches in separate parameter sets? I'd certainly rather have it be obvious that they can't be used together than to allow them together but ignore one (or worse, produce runtime errors if they're used together)...

-

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 6 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 5 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 4 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: > I think you mean that it's very unintuitive 😕 

Under "intuitive" I mean that we use the cmdlet 10 years "as is" without many complaints (In other words, users get results they expect). 
All completely different if you look on the inner side of the cmdlet. I guess it was one of the first cmdlets of Powershell. Today such a design would be immediately rejected. The code is tricky and any chip change results in multiple test crush.
And we absolutely cannot add parameter sets (and moreover we cannot add new exceptions) because this is _breaking change_. 
As I said above, time to replace this cmdlet to a new one with stringent modern design, better UX  and better performance.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw I fix breaking change and add your test.

**Comment by iSazonov** [comment]
@RichardSiddaway Sorry, this is the right decision but it is impossible to add parameter sets w/o breaking change. See my comment above.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 1 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: > I think @PowerShellTeam needs to mark this moment in documentation: "make more comments in your code".

I absolutely agree! I will update [coding-conventions](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#coding-conventions) to include this.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs [hunk 7]
```diff
@@ -334,9 +334,11 @@ private void ProcessExpressionParameter()
                 _expandMshParameterList = processor.ProcessParameters(new string[] { ExpandProperty }, invocationContext);
             }
 
-            if (ExcludeProperty != null)
+            // ExpandProperty skip processing ExcludeProperty
+            if (ExcludeProperty != null && string.IsNullOrEmpty(ExpandProperty))
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: We cannot ignore `-ExcludeProperty` just because `-ExpandProperty` is specified. See the script below:
```
PS:88> $s = Get-Process -Id $pid | Select-Object -Property Process* -ExpandProperty Modules
PS:89> $s[0].ProcessName
powershell
PS:90> $s[0].ProcessorAffinity
255
PS:91> $p = Get-Process -Id $pid | Select-Object -Property Process* -ExcludeProperty ProcessorAffinity
-ExpandProperty Modules
PS:92> $p[0].ProcessName
powershell
PS:93> $p[0].ProcessorAffinity
PS:94>
```
`-ExcludeProperty` took effect and excluded the property `ProcessorAffinity`, so only `ProcessName` got added to each Module object.

I feel it's OK to break the `Select-Object -ExcludeProperty value1 -ExpandProperty value2` usage based on my analysis of the usage of `Select-Object` in GitHub. Please see my comment above.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 6 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 5 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 4 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 1 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: #Closed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs [hunk 8]
```diff
@@ -419,18 +421,10 @@ private void ProcessParameter(MshParameter p, PSObject inputObject, List<PSNoteP
                 }
             }
 
+            // allow Select-Object -Property noexist-name to return a PSObject with property noexist-name,
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: It would be clearer if you add quotes around `Select-Object -Property noexist-name`.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: I analyzed the usages of `Select-Object` in GitHub to find the usages like `Select-Object -ExcludeProperty ProcessName -ExpandProperty Modules`, which might be broken by this fix. The analysis scripts and results can be found [here](https://gist.github.com/daxian-dbw/07847ff5ef6da2c100e8b22bd94f0feb).

1074 .ps1/.psm1 files were found in GitHub with `Select-Object` in them, and totally 3376 instances of `Select-Object` were found. Out of them, 2 usages might be broken by the fix on `-ExcludeProperty`:
```
## FIND 2 usages like 'Select-Object -ExcludeProperty value -ExpandProperty value'
PS:96> $results | ? BreakingCase

UsesSelectObject     : True
FileName             : F:\temp\selectobject-dir-psm1\ECSNamespace.psm1
NeedManualInspection : False
BreakingCase         : True
Line                 : 48

UsesSelectObject     : True
FileName             : F:\temp\selectobject-dir-psm1\ECSvdc.psm1
NeedManualInspection : False
BreakingCase         : True
Line                 : 158
```
[ECSNamespace.psm1](https://raw.githubusercontent.com/bottkars/ECShell/6b2531a84e5107ccb1c6ab71a18e766f057efa14/Multitenancy/ECSNamespace.psm1) has `Select-Object -ExpandProperty $Expandproperty -ExcludeProperty $Excludeproperties` in it:
```
Invoke-RestMethod -Uri $Uri -Headers $ECSAuthHeaders -Method Get  -ContentType $ContentType | Select-Object -ExpandProperty $Expandproperty -ExcludeProperty $Excludeproperties | Select-Object $Properties #,@{N="NamespaceID";E={$_.id}}
```
[ECSvdc.psm1](https://raw.githubusercontent.com/bottkars/ECShell/6b2531a84e5107ccb1c6ab71a18e766f057efa14/Provisioning/ECSvdc.psm1) has `Select-Object -ExpandProperty vdc -ExcludeProperty $Excludeproperty` in it:
```
Invoke-RestMethod -Uri $Uri -Headers $ECSAuthHeaders -Method $Method -ContentType $ContentType  | Select-Object -ExpandProperty $Expandproperty | Select-Object -ExpandProperty vdc -ExcludeProperty $Excludeproperty
```

Although we actually find the usages that might be broken by this fix, but those 2 usages seems wrong in the first place -- the author indented to exclude some properties, which actually doesn't happen. 

Given the data, I feel it's OK to have this "breaking change" introduced by this fix.

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: To avoid this breaking change, we will have to do something like the following:
```
When -ExcludeProperty is specified
    #1 if -Property is specified
        always respect -ExcludeProperty
    #2 else if -ExpandProperty is specified
        Ignore -ExcludeProperty
    #3 else
        set Property to "*" and respect -ExcludeProperty
```
But this will again raise confusion about how `-ExcludeProperty` actually works (when and when not it takes effect), and probably someone will see `#2` as a bug too. Therefore, I'm more inclined to accept always having `#3` when `-Property` is not specified, though it would introduce the breaking change discussed above. 
@lzybkr @adityapatwardhan @vors, what are your opinions on this?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 7 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Ok, waiting team approve above then revert the change.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 8 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Done.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 7 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: While waiting I made refactoring. See comment above.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw I made refactoring ProcessExpressionParameter() to the logic, add your test, add comments.

Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: I'd be in favour of non-breaking change, so I agree with pseudo-code @daxian-dbw proposed.
Although it's still far from the clean design, it's a) non-breaking for existing legit usages b) better than what we currently have. So I'm considering it a reasonable step in the right direction.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @vors Thanks!
@daxian-dbw I have implemented your suggestion in the last commit. Waiting for your conclusion.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw Thanks for summing-up. Await @lzybkr.
And my question remains whether to announce this cmdlet depredicated and design a new cmdlet with simple clear logic and easy to support code?

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @lzybkr this discussion thread was auto-collapsed in the main PR page, but you can read it at https://github.com/PowerShell/PowerShell/pull/2421/files/70c30384c74d9f6392b13cc9d04baf9eb5f8c364

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw @lzybkr your conclusion?

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Sorry @iSazonov for the long delay.
@lzybkr and I had a quick discussion about this, and I think @lzybkr agreed that we should take the original fix with the *unlikely* breaking change.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw  Thanks! Good news!
Only I seem to be lost. What are my next steps? 
It seems I made all the changes you want in the code.

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: The conclusion is to take your changes to `ProcessExpressionParameter()` in your second commit -- always add "*" to `-Property` when `-ExcludeProperty` is specified but `-Property` is not : https://github.com/PowerShell/PowerShell/pull/2421/commits/70c30384c74d9f6392b13cc9d04baf9eb5f8c364

I think you implemented the pseudo code I listed above to avoid the breaking change, but as we discussed here, that will raise new confusion about how `-ExlcudeProperty` works. So I believe it's reasonable to take your original changes to `ProcessExpressionParameter()` to make it simple for users to understand.

Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Note that I hadn't actually concluded anything after my discussion w/ @daxian-dbw, but after some of my own experiments, I do agree. The existing behavior is surprising and not useful, anyone potentially hit by the breaking change wasn't getting what they wanted anyway.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: @daxian-dbw @lzybkr Thanks for the clarification and the final conclusion! I'll make the rollback.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 7 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Done.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Done.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 9]
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+
+        # Issue #2351
+	It "Select-Object with explicit Property = '*' exclude single property"{
+		$results = [pscustomobject]@{Thing="thing1"} | Select-Object -Property * -ExcludeProperty thing
+		$results.psobject.Properties.Item("Thing") | Should Be $null
+		$results.psobject.Properties.Item("*") | Should Be $null
+	}
+
+        # Issue #2351
+	It "Select-Object with implicit Property = '*' exclude not single prperty"{
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: > prperty

A typo, should be 'property'

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 10]
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -ExcludeProperty Param2 
+		$results.Param2 | Should Be $null
+		$results.Thing | Should Be "thing1"
+	}
+
+        # Issue #2351
+	It "Select-Object with explicit Property = '*' exclude not single prperty"{
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: > prperty

Same typo :)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 [hunk 11]
```diff
@@ -248,3 +248,51 @@ Describe "Select-Object DRT basic functionality" -Tags "CI" {
...
+		$results = [pscustomobject]@{Thing="thing1";Param2="param2"} | Select-Object -Property * -ExcludeProperty Param2 
+		$results.Param2 | Should Be $null
+		$results.Thing | Should Be "thing1"
+	}
+
+    It "Select-Object with ExpandProperty and Property don't skip processing ExcludeProperty" {
+		$p = Get-Process -Id $pid | Select-Object -Property Process* -ExcludeProperty ProcessorAffinity -ExpandProperty Modules
+		$p[0].psobject.Properties.Item("ProcessorAffinity") | Should Be $null
+    }
+}
```
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Can you please add a new line?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 3 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Thank you @iSazonov 
#Closed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 7 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: #Closed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs (hunk 8 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: #Closed

**Comment by daxian-dbw** [comment]
Thanks @iSazonov for the updates! Once the typos in tests are fixed, the PR will be ready to merge.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 9 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 10 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: It's fate :)
Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 11 above)
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Fixed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 9 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 10 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: #Closed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1 (hunk 11 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-Object.Tests.ps1]: Thanks. #Closed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs [hunk 12]
```diff
@@ -337,7 +339,13 @@ private void ProcessExpressionParameter()
             if (ExcludeProperty != null)
             {
                 _exclusionFilter = new MshExpressionFilter(ExcludeProperty);
+                // ExcludeProperty implies -Property * for better UX
+                if ((Property == null) || (Property.Length == 0))
+                 {
+                    Property = new Object[]{"*"};
+                    _propertyMshParameterList = processor.ProcessParameters(Property, invocationContext);                }
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/select-object.cs]: Done.

**Review by daxian-dbw — APPROVED** [review]
LGTM.

**Comment by iSazonov** [comment]
@daxian-dbw Thanks very much!

**Comment by daxian-dbw** [comment]
@iSazonov Thank you for your patience and the great contribution 👍
